### DISCUSSION TITLE
Error: ❌ Failed to prepare provider: Unable to dereference or decycle OpenAPI specification.

### DISCUSSION THREAD

**Discussion by dsteele101** [superfaceai/.github#7] [Q&A]
When trying to run superface, I get the error as below:
`superface prepare openapi-updated.json "amelia"
⠇ info - Ingesting: "{
  "openapi": "3.0.1",
  "info": {
    "title": "..."
 ›   Error: ❌ Failed to prepare provider: Unable to dereference or decycle OpenAPI specification.`

I have converted Swagger 2 spec to OpenAPI 3.0.1.

Here is the Swagger 2 spec: https://github.com/dsteele101/superface/blob/main/amelia_swagger.json

And the OpenAI 3.0.1 spec: https://github.com/dsteele101/superface/blob/main/openapi.json

I tried to remove circular references here: https://github.com/dsteele101/superface/blob/main/openapi-updated.json

Any insights into how I resolve this error?

**Comment by jnv** [comment]
I second this one, I ran into the same exact issue with [OAS for Mews](https://api.mews.com/Swagger/connector/swagger.yaml). I have yet to try linting it though.

**Comment by zdne** [comment]
@dsteele101 we are looking into it and will keep you posted cc @Jakub-Vacek

**Comment by Jakub-Vacek** [comment]
@dsteele101 we are still looking into it. I will let you know as soon as this is fixed. 
Also, thanks for trying Superface.

**Reply by dsteele101** [reply]
Re-tested on 4.0.1 with the same results.

**Reply by jnv** [reply]
I think v4.0.1 doesn't include any fix for this.

@Jakub-Vacek any updates?

**Reply by dsteele101** [reply]
Was not expected, but thought I would try anyway, as the testing is quick.

**Reply by Jakub-Vacek** [reply]
I am working on PR, there are some aspects that we need to think through so it will take some time.

**Comment by zdne** [comment]
@Jakub-Vacek any updates here?

**Reply by Jakub-Vacek** [reply]
I am half through the PR. I have tested some possible ways to fix this, now I am implementing it for OAS v.2 and OAS v.3.

**Comment by zdne** [comment]
FWIW I've ran into the same problem with Stripe here:

```
zdenek@razorback qa % sf prepare https://raw.githubusercontent.com/stripe/openapi/master/openapi/spec3.yaml "stripe"
⠏ info - Ingesting: "https://raw.githubusercontent.com/stripe/openapi/m..."
 ›   Error: ❌ Failed to prepare provider: Unable to dereference or decycle OpenAPI specification.
```

**Reply by dsteele101** [reply]
Came across this issue with the latest version [here](https://resources.docs.salesforce.com/rel1/doc/en-us/static/misc/iot-connect-api-swagger-45.0.yaml)

superface prepare https://resources.docs.salesforce.com/rel1/doc/en-us/static/misc/iot-connect-api-swagger-45.0.yaml "sf-iot"
⠴ info - Ingesting: "https://resources.docs.salesforce.com/rel1/doc/en-..."
 ›   Error: ❌ Failed to prepare provider: Unable to dereference or decycle OpenAPI specification.Failed to prepare provider: Unable to dereference or decycle OpenAPI specification.
 ›
 ›   If you need help with this error, please raise an issue on GitHub Discussions: https://sfc.is/discussions
 ›
 ›   Please include the following information:
 ›   - command you ran
 ›   - error message
 ›   - version of the CLI, OS: superface cli/4.1.1 (darwin-arm64) node-v20.1.0

superface version
@superfaceai/cli/4.1.1 darwin-arm64 node-v20.1.0

**Comment by Jakub-Vacek** [comment]
Hi @dsteele101. I am finishing the PR fixing this issue. It will be released in next few days. I will let you know as soon as it is released :)

**Reply by Jakub-Vacek** [reply]
@dsteele101 Fix is released on production,

**Reply by dsteele101** [reply]
Tested with 4.1.2

superface prepare https://resources.docs.salesforce.com/rel1/doc/en-us/static/misc/iot-connect-api-swagger-45.0.yaml "sf-iot"
⠏ Preparing provider definition
 ›   Error: ❌ Failed to prepare provider: Unable to dereference OpenAPI specification. Detail: MissingPointerError: Token "DirectoryItemCollectionRepresentation" does not exist.
 ›
 ›   If you need help with this error, please raise an issue on GitHub Discussions: https://sfc.is/discussions
 ›
 ›   Please include the following information:
 ›   - command you ran
 ›   - error message
 ›   - version of the CLI, OS: superface cli/4.1.2 (darwin-arm64) node-v20.1.0

**Reply by dsteele101** [reply]
It did work with my original API spec, so perhaps the error I posted above is something else altogether.

**Reply by Jakub-Vacek** [reply]
Yes, it looks like problem with salesforce OAS Specification: there is a reference to `DirectoryItemCollectionRepresentation` but there is no actual definition schema of that. If you comment out invalid reference (or provide correct schema) it works.

**Comment by Jonneal3** [comment]
<img width="1047" alt="Screenshot 2023-10-12 at 10 12 17 PM" src="https://github.com/superfaceai/.github/assets/92951162/b2559c92-9272-494e-bd64-f8ecb890e33c">

Wont prepare a teh basic google calendar URL-- thoughts?
